src/openssl/ssl/socket.cr

```diff
@@ -49,11 +49,15 @@ abstract class OpenSSL::SSL::Socket < IO
   class Server < Socket
     def initialize(io, context : Context::Server = Context::Server.new, sync_close : Bool = false)
       super(io, context, sync_close)
-
-      ret = LibSSL.ssl_accept(@ssl)
-      unless ret == 1
-        io.close if sync_close
-        raise OpenSSL::SSL::Error.new(@ssl, ret, "SSL_accept")
+      begin
+        ret = LibSSL.ssl_accept(@ssl)
+        unless ret == 1
+          io.close if sync_close
+          raise OpenSSL::SSL::Error.new(@ssl, ret, "SSL_accept")
+        end
+      rescue ex
```

Comment:
The exception caught here is from `io.close` or the `raise OpenSSL::SSL::Error`?
If it's for the later then the `finalize` call could be put between creating the exception and the raise

```cr
ex = OpenSSL::SSL::Error.new(@ssl, ret, "SSL_accept")
finalize # or even LibSSL.ssl_free(@ssl) directly.
raise ex
```

Reply:
Can io.close raise an exception?  I also forgot to add the client  equivalent but I have added it now. The client kind of feels like it needs an ensure clause.  Should I keep the same pattern for both do you think?  I just thought it'd be easier to maintain since it's more obvious that intermediate code goes "in the ensure/catch block" for the server side, it follows a pattern, etc.  Thanks.

Reply:
Ensure would run even if everything goes right. I'm pretty sure you don't want to free the socket then.

Reply:
`io.close`can theoretically raise. But I don't think that matters here because when `ret == 1` you need to free the SSL socket anyway.
So I think this change could just be a call to `LibSSL.ssl_free` right before the `io.close`.

Reply:
OK how do I remove a commit and replace it in these github review PR's, do the fixes and create a new PR?  Add another commit?  Thanks.

Reply:
Add another commit, we can squash on merge. No worries :-)

Reply:
OK I tried putting `LibSSL.ssl_free` right before `io.close`.  It segfaults on the line after that `raise OpenSSL::SSL::Error.new(@ssl, ret, "SSL_accept")` My hunch is because the ssl_free call right before that was actually successful, so it lost error context...hmm..so I guess we could reorder everything there or use the rescue block...

Reply:
If you do

```cr
ex = OpenSSL::SSL::Error.new(@ssl, ret, "SSL_accept")
LibSSL.ssl_free(@ssl)
raise ex
```

It should let you create the exception, free and _then_ raise. I’m assuming `@ssl` is not used after the exception is initialized.